src/libsystemd-network/sd-dhcp-server.c

```diff
@@ -712,8 +715,70 @@ static int get_pool_offset(sd_dhcp_server *server, be32_t requested_ip) {
         return be32toh(requested_ip & ~server->netmask) - server->pool_offset;
 }
 
-static int dhcp_server_relay_message(sd_dhcp_server *server, DHCPMessage *message, size_t opt_length) {
+static size_t relay_option_length(sd_dhcp_server *server) {
+        size_t length = 0;
+
+        assert(server);
+
+        if (server->agent_circuit_id)
+                length += strlen(server->agent_circuit_id) + 2;
+        if (server->agent_remote_id)
+                length += strlen(server->agent_remote_id) + 2;
+
+        return length;
+}
+
+static uint8_t *append_suboption(uint8_t *dst, size_t dst_len, uint8_t subcode, const char *value, size_t length) {
+        assert(dst);
+        assert(value);
+
+        if (dst_len < length + 2)
+                return NULL;
+        dst[0] = subcode;
+        dst[1] = length;
+        memcpy(dst + 2, value, length);
+        return dst + 2 + length;
+}
```

Comment:
This is mostly equivalent to `option_append()` in dhcp-option.c. Maybe, it is better to reuse that.

Reply:
Ok, I reused that, but I have not called option_append for appending suboptions directly, because option_append handles a lot of options specially and Ids of DHCP options and RFC3046 
 suboptions are intersecting.

Reply:
Ok, I refactored this code, think it's much better now. No double-copy, and the option appending logic is in the option_append together with logic of appending of other options. Please check and tell me what do you think.